providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java

```diff
@@ -109,9 +106,32 @@ public FlagdProvider(final FlagdOptions options) {
         this.maxEventStreamRetries = maxEventStreamRetries;
         this.cache = new FlagdCache(cache, maxCacheSize);
         this.eventStreamAliveSync = new Object();
+    }
+
+    @Override
+    public void initialize(EvaluationContext evaluationContext) {
+        this.serviceBlockingStub = this.serviceBlockingStub
+                .withWaitForReady()
+                .withDeadline(Deadline.after(30, TimeUnit.SECONDS));
+        this.serviceStub = this.serviceStub
+                .withWaitForReady()
+                .withDeadline(Deadline.after(30, TimeUnit.SECONDS));
```

Comment:
I'm a bit confused with how this works in conjunction with `restartEventStream` below, which is called by the `EventStreamObserver` when there's an error. As `withWaitForReady` explains:

```This may dramatically increase the latency of the RPC, but avoids failing "unnecessarily." The default queues the RPC until an attempt to connect has completed, but fails RPCs without sending them if unable to connect.```

I'm concerned the 30s timeout _may_ be conflicting with our custom retry logic. If I'm understanding everything here correctly, it means that if we're having trouble connecting entirely, we won't see an error until 30s later. I think the way our retry works, it may be better to use the default method, since it will then retry using our backoff (because an error will be thrown).

This brings up another point... which we may want to address in another PR: we don't have integration tests that fully test our retry logic, as far as I can tell.

Reply:
OK so I confirmed this locally. _Without_ the `withWaitForReady` and `withDeadline` stuff, we attempt our normal reconnect protocol immediately if flagd is unreachable. _with it_, we wait 30s until we do. I think we should remove those.

Reply:
Agree with the wait time argument

Also, for flagd provider, connection with backend can only be detected with the event stream connectivity. So the provider readiness can be blocked till we observe a notification on `eventStreamAliveSync` (ex: `this.eventStreamAliveSync.wait();`)

However, we might need to consider running event handling concurrently 🤔  right now it's done in the same thread and also have thread sleeps [^1].

[^1]: https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java#L123

Reply:
@Kavindu-Dodan might have a good point here.

Reply:
I added a deadline because I was following this documentation of gRPC: https://grpc.io/blog/deadlines/
The `withWaitForReady` delegate the gRPC library to do the heavy lifting in making sure the connection is established instead of having to do it by ourselves. 

The code needs some refactoring since the data- and control-flow are quite hard to follow. If we want to implement a blocking initialize method, I can only think about two quick solutions without many code changes:
- we set the deadline in an exp way to match the backoff that we have
- we create a thread that makes the connection and waits for the `this.eventStreamAliveSync.wait();` and we join on it 

I think the more clean implementation is the first option, but I am open to your input @toddbaert and @Kavindu-Dodan  :)

Reply:
Just to be clear, the SDK itself runs the `initialize` in a thread it manages. This was intentional to prevent initialization from ever blocking the whole application.

> - we set the deadline in an exp way to match the backoff that we have (removing the thread sleep)
> - we create a thread that makes the connection and waits for the this.eventStreamAliveSync.wait(); and we join on it

Personally I'm completely fine with either of these, as long as we conform to the requirements (exponential back-off with a max-delay and max attempts).